### DISCUSSION TITLE
Cannot compile Firefox for whatever reason.

### DISCUSSION THREAD

**Discussion by ghost** [rust9x/rust#28] [Q&A] [answered]
Hello, there! I've been trying to compile Luminafox (my Firefox fork targeting Windows XP, Vista, 7 & 8.1) using rust9x, but sadly it refuses to compile and throws out an error in the Rust toolchain checks. Any fix would be appreciated.
![image](https://github.com/rust9x/rust/assets/167991948/808b145a-f981-466a-a5f5-4048f2e87082)

**Comment by yuztra** [comment]
for clarification: the dev(s) forgot to include cargo in rust9x, so please include it in the next update of rust9x.

**Comment by seritools** [comment, marked as answer]
There might be the misconception that just because rust9x exists all Rust crates (e.g. in Firefox) will compile without issue. Many crates in general, and also for Firefox in particular, might use their own imports of different Windows APIs, which might make them incompatible to any given older Windows version, and might need additional forks or fixes.

Cargo itself is one such project, and won't be supported on the legacy systems.

Shipping cargo is also not necessary for regular Rust usage with rust9x. This sounds more like a limitation of the Firefox build system (not expecting to work with custom toolchains), not of rust9x. If you want to try building/shipping cargo with rust9x, feel free to enable cargo in https://github.com/rust9x/rust/blob/rust9x/config.rust9x.toml#L22 and recompile, and let me know if it works.

**Reply by ErichDonGubler** [reply]
In case it helps, Firefox currently brings in a copy of the `windows` 0.52.0 crate for building. It avoids actually vendoring in the source, but should be perfectly patchable with work: https://searchfox.org/mozilla-central/rev/8ec3cc0472ad4f51b254728d024b696eaba82ba0/build/rust/windows/Cargo.toml
